ProcessLib/RichardsFlow/RichardsFlowFEM.h

```diff
@@ -92,75 +138,80 @@ class LocalAssemblerData : public RichardsFlowLocalAssemblerInterface
         pos.setElementID(_element.getID());
+        const int material_id =
+            _process_data.material->getMaterialID(pos.getElementID().get());
+        const Eigen::MatrixXd& perm = _process_data.material->getPermeability(
+            material_id, t, pos, _element.getDimension());
+        assert(perm.rows() == _element.getDimension() || perm.rows() == 1);
+        GlobalDimMatrixType permeability = GlobalDimMatrixType::Zero(
+            _element.getDimension(), _element.getDimension());
+        if (perm.rows() == _element.getDimension())
+            permeability = perm;
+        else if (perm.rows() == 1)
+            permeability.diagonal().setConstant(perm(0, 0));
 
         for (unsigned ip = 0; ip < n_integration_points; ip++)
         {
             pos.setIntegrationPoint(ip);
-            auto const& sm = _shape_matrices[ip];
-            auto const& wp = _integration_method.getWeightedPoint(ip);
-
-            double P_int_pt = 0.0;
-            NumLib::shapeFunctionInterpolate(local_x, sm.N, P_int_pt);
-
-            auto const K = _process_data.intrinsic_permeability(t, pos)[0];
-            auto const poro = _process_data.porosity(t, pos)[0];
-            auto const mu = _process_data.viscosity(t, pos)[0];
-            auto const storage = _process_data.storage(t, pos)[0];
-            double const Pc = -P_int_pt;
-
-            double dSwdPc = interpolated_Pc.getDerivative(Pc);
-            if (Pc > interpolated_Pc.getSupportMax())
-                dSwdPc = interpolated_Pc.getDerivative(
-                    interpolated_Pc.getSupportMax());
-            else if (Pc < interpolated_Pc.getSupportMin())
-                dSwdPc = interpolated_Pc.getDerivative(
-                    interpolated_Pc.getSupportMin());
-
-            double const Sw = interpolated_Pc.getValue(Pc);
+            double p_int_pt = 0.0;
+            NumLib::shapeFunctionInterpolate(local_x, _ip_data[ip].N, p_int_pt);
+            const double& temperature = _process_data.temperature(t, pos)[0];
+            auto const porosity = _process_data.material->getPorosity(
+                material_id, t, pos, p_int_pt, temperature, 0);
+
+            double const pc_int_pt = -p_int_pt;
+
+            double const Sw = _process_data.material->getSaturation(
+                material_id, t, pos, p_int_pt, temperature, pc_int_pt);
             _saturation[ip] = Sw;
 
-            double const k_rel = interpolated_Kr.getValue(Sw);
+            double const dSw_dpc =
+                _process_data.material->getSaturationDerivative(
+                    material_id, t, pos, p_int_pt, temperature, Sw);
+
             // \TODO Extend to pressure dependent density.
             double const drhow_dp(0.0);
 
             double const mass_mat_coeff =
-                storage * Sw + poro * Sw * drhow_dp - poro * dSwdPc;
-            double const K_mat_coeff = K * k_rel / mu;
+                porosity * Sw * drhow_dp - porosity * dSw_dpc;
```

Comment:
Compared to the previous version the term `storage * Sw` is missing. Is the new implementation correct?